lib/sda/node.go

```diff
@@ -0,0 +1,302 @@
+func (n *Node) IsRoot() bool {
+	tn, ok := n.TreeNode()
+	if !ok {
+		panic("IsRoot() called TreeNode(): did not find ourself: it's bad")
+	}
+	return tn.Parent == nil
+}
+
+// IsLeaf returns whether whether we are at the bottom of the tree
+func (n *Node) IsLeaf() bool {
+	tn, ok := n.TreeNode()
+	if !ok {
+		panic("IsLeaf() called TreeNode(): did not find ourself: it's bad")
+	}
+	return len(tn.Children) == 0
+}
+
+// SendTo sends to a given node
+func (n *Node) SendTo(to *TreeNode, msg interface{}) error {
+	if to == nil {
+		return errors.New("Sent to a nil TreeNode")
+	}
+	return n.overlay.SendToTreeNode(n.token, to, msg)
+}
+
+// Tree returns the tree of that node
+func (n *Node) Tree() *Tree {
+	return n.overlay.TreeFromToken(n.token)
+}
+
+// EntityList returns the entity-list
+func (n *Node) EntityList() *EntityList {
+	return n.Tree().EntityList
+}
+
+// RegisterChannel takes a channel with a struct that contains two
+// elements: a TreeNode and a message. It will send every message that are the
+// same type to this channel. NOTE: In legacy network library, you had to
+// register the message type to the network library. Using this function you
+// don't have to anymore, it will automatically do the registration for you.
+func (n *Node) RegisterChannel(c interface{}) error {
+	cr := reflect.TypeOf(c)
+	// Check we have the correct channel-type
+	if cr.Kind() != reflect.Chan {
+		return errors.New("Input is not channel")
+	}
+	if cr.Elem().Kind() != reflect.Struct {
+		return errors.New("Input is not channel of structure")
+	}
+	if cr.Elem().NumField() != 2 {
+		return errors.New("Input is not channel of structure with 2 elements")
+	}
+	dbg.Lvl3(cr.Elem().Field(0).Type)
+	if cr.Elem().Field(0).Type != reflect.TypeOf(TreeNode{}) {
+		return errors.New("Input-channel doesn't have TreeNode as element")
+	}
+	// Automatic registration of the message to the network library.
+	typ := network.RegisterMessageUUID(network.RTypeToUUID(cr.Elem().Field(1).Type),
+		cr.Elem().Field(1).Type)
+	n.channels[typ] = c
+	dbg.Lvl3("Registered channel", typ)
+	return nil
+}
+
+// ProtocolInstance returns the instance of the running protocol
+func (n *Node) ProtocolInstance() ProtocolInstance {
+	return n.instance
+}
+
+// ProtocolInstantiate creates a new instance of a protocol given by it's name
+func (n *Node) protocolInstantiate() error {
+	if n.token == nil {
+		return errors.New("Hope this is running in test-mode")
+	}
+	p, ok := protocols[n.token.ProtocolID]
+	if !ok {
+		return errors.New("Protocol doesn't exist")
+	}
+	tree := n.overlay.Tree(n.token.TreeID)
+	if tree == nil {
+		return errors.New("Tree does not exists")
+	}
+	if n.overlay.EntityList(n.token.EntityListID) == nil {
+		return errors.New("EntityList does not exists")
+	}
+	_, ok = n.TreeNode()
+	if !ok {
+		return errors.New("We are not represented in the tree")
+	}
+	n.instance = p(n)
+	return nil
+}
+
+func (n *Node) DispatchFunction(msg []*SDAData) error {
```

Comment:
All of these "Dispatch*****" methods should be non exported. Users should be doing anything with that.